packages-ts/gauntlet-terra-contracts/src/lib/contracts.ts

```diff
@@ -28,109 +29,164 @@ export type TerraABI = {
 
-const defaultContractVersions = {
-  [CONTRACT_LIST.FLAGS]: DEFAULT_RELEASE_VERSION,
-  [CONTRACT_LIST.DEVIATION_FLAGGING_VALIDATOR]: DEFAULT_RELEASE_VERSION,
-  [CONTRACT_LIST.OCR_2]: DEFAULT_RELEASE_VERSION,
-  [CONTRACT_LIST.ACCESS_CONTROLLER]: DEFAULT_RELEASE_VERSION,
-  [CONTRACT_LIST.CW20_BASE]: DEFAULT_CWPLUS_VERSION,
-  [CONTRACT_LIST.CW4_GROUP]: DEFAULT_CWPLUS_VERSION,
-  [CONTRACT_LIST.MULTISIG]: DEFAULT_CWPLUS_VERSION,
-}
-export const getContractABI = (contractId: CONTRACT_LIST): TerraABI => {
-  // Possible paths depending on how/where gauntlet is being executed
-  const possibleContractPaths = [
-    path.join(__dirname, './artifacts/contracts'),
-    path.join(process.cwd(), './contracts'),
-    path.join(process.cwd(), '../../contracts'),
-    path.join(process.cwd(), './packages-ts/gauntlet-terra-contracts/artifacts/contracts'),
-  ]
-
-  const toDirName = (contractId: CONTRACT_LIST) => contractDirName[contractId]
-
-  const abi = possibleContractPaths
-    .filter((path) => existsSync(`${path}/${toDirName(contractId)}/schema`))
-    .map((contractPath) => {
-      const toPath = (type) => {
-        if (contractId == CONTRACT_LIST.CW20_BASE && type == 'execute_msg') {
-          return path.join(contractPath, `./${toDirName(contractId)}/schema/cw20_${type}`)
-        } else {
-          return path.join(contractPath, `./${toDirName(contractId)}/schema/${type}`)
-        }
-      }
-      return {
-        execute: io.readJSON(toPath('execute_msg')),
-        query: io.readJSON(toPath('query_msg')),
-        instantiate: io.readJSON(toPath('instantiate_msg')),
-      }
-    })
-  if (abi.length === 0) {
-    logger.error(`ABI not found for contract ${contractId}`)
+  constructor(id, dirName, defaultVersion = DEFAULT_RELEASE_VERSION) {
+    super(id, dirName, defaultVersion)
   }
+}
+
+class CosmWasmContract extends Contract {
+  readonly downloadUrl = `https://github.com/CosmWasm/cw-plus/releases/download/`
 
-  return abi[0]
+  constructor(id, dirName, defaultVersion = DEFAULT_CWPLUS_VERSION) {
+    super(id, dirName, defaultVersion)
+  }
 }
 
-export const getContract = async (id: CONTRACT_LIST, version): Promise<Contract> => {
-  version = version ? version : defaultContractVersions[id]
-  // Preload contracts
-  return {
-    id,
-    abi: getContractABI(id),
-    bytecode: await getContractCode(id, version),
+class Contracts {
+  async getContract(id: CONTRACT_LIST, version: string): Promise<Contract> {
+    if (this[id]) {
+      return await this[id].getContract(version)
+    } else {
+      throw new Error(`Contract ${id} not found!`)
+    }
+  }
+  addChainlink = (id: CONTRACT_LIST, dirName: string) => {
+    this[id] = new ChainlinkContract(id, dirName)
+    return this
+  }
+  addCosmwasm = (id: CONTRACT_LIST, dirName: string) => {
+    this[id] = new CosmWasmContract(id, dirName)
+    return this
   }
 }
+
+export const contracts = new Contracts()
```

Comment:
What's the benefit of this new structure? Why not just adding `addresses` to the old `Contract` type?

Reply:
The number of different global data structures used to define different properties of the same contract was sprawling and scattered throughout `contracts.ts`.  Accessing two different properties of a contract required importing several different data structures.  Making `Contract` into a legitimate class and keeping all of the properties of each contract in there together was the main aim of this refactor, which I think will help a lot with readability.

For example, instead of:
```
-const contractDirName = {
-  [CONTRACT_LIST.FLAGS]: 'flags',
-  [CONTRACT_LIST.DEVIATION_FLAGGING_VALIDATOR]: 'deviation-flagging-validator',
-  [CONTRACT_LIST.OCR_2]: 'ocr2',
-  [CONTRACT_LIST.PROXY_OCR_2]: 'proxy-ocr2',
-  [CONTRACT_LIST.ACCESS_CONTROLLER]: 'access-controller',
-  [CONTRACT_LIST.CW20_BASE]: 'cw20_base',
-  [CONTRACT_LIST.CW4_GROUP]: 'cw4_group',
-  [CONTRACT_LIST.MULTISIG]: 'cw3_flex_multisig',
-}
 
-const defaultContractVersions = {
-  [CONTRACT_LIST.FLAGS]: DEFAULT_RELEASE_VERSION,
-  [CONTRACT_LIST.DEVIATION_FLAGGING_VALIDATOR]: DEFAULT_RELEASE_VERSION,
-  [CONTRACT_LIST.OCR_2]: DEFAULT_RELEASE_VERSION,
-  [CONTRACT_LIST.ACCESS_CONTROLLER]: DEFAULT_RELEASE_VERSION,
-  [CONTRACT_LIST.CW20_BASE]: DEFAULT_CWPLUS_VERSION,
-  [CONTRACT_LIST.CW4_GROUP]: DEFAULT_CWPLUS_VERSION,
-  [CONTRACT_LIST.MULTISIG]: DEFAULT_CWPLUS_VERSION,
-}
```
combined with special case hacks in specific functions, like this:
```
-    switch (contractId) {
-      case CONTRACT_LIST.CW20_BASE:
-      case CONTRACT_LIST.CW4_GROUP:
-      case CONTRACT_LIST.MULTISIG:
-        url = `https://github.com/CosmWasm/cw-plus/releases/download/${version}/${contractId}.wasm`
-        break
-      default:
-        url = `https://github.com/smartcontractkit/chainlink-terra/releases/download/${version}/${contractId}.wasm`
```
We have:
```
export abstract class Contract {
  // Contract metadata, initialized in constructor
  readonly id: CONTRACT_LIST
  readonly defaultVersion: string
  readonly dirName: string
  readonly downloadUrl: string

  // Only load bytecode & schema later if needed
  abi: TerraABI
  bytecode: string
  addresses: AccAddress[] = []
  address: AccAddress // first address, for convenience

  constructor(id, dirName, defaultVersion) {
    this.id = id
    this.defaultVersion = defaultVersion
    this.dirName = dirName
  }
...
}
...
class ChainlinkContract extends Contract {
  readonly downloadUrl = `https://github.com/smartcontractkit/chainlink-terra/releases/download/`

  constructor(id, dirName, defaultVersion = DEFAULT_RELEASE_VERSION) {
    super(id, dirName, defaultVersion)
  }
}

class CosmWasmContract extends Contract {
  readonly downloadUrl = `https://github.com/CosmWasm/cw-plus/releases/download/`

  constructor(id, dirName, defaultVersion = DEFAULT_CWPLUS_VERSION) {
    super(id, dirName, defaultVersion)
  }
}
...
export const contracts = new Contracts()
  .addChainlink(CONTRACT_LIST.FLAGS, 'flags')
  .addChainlink(CONTRACT_LIST.DEVIATION_FLAGGING_VALIDATOR, 'deviation-flagging-validator')
  .addChainlink(CONTRACT_LIST.OCR_2, 'ocr2')
  .addChainlink(CONTRACT_LIST.PROXY_OCR_2, 'proxy-ocr2')
  .addChainlink(CONTRACT_LIST.ACCESS_CONTROLLER, 'access-controller')
  .addCosmwasm(CONTRACT_LIST.CW20_BASE, 'cw20_base')
  .addCosmwasm(CONTRACT_LIST.CW4_GROUP, 'cw4_group')
  .addCosmwasm(CONTRACT_LIST.MULTISIG, 'cw3_flex_multisig')
```
The `Contracts` class is less important, at present it's just a collection of 3 small convenience functions, two for defining a new `Contract` (`addCosmwasm` and `addChainlink`) and one for accessing a `Contract`.  They were added to slightly reduce the verbosity of the lines of code we have for defining and accessing `Contracts`.  These could probably just as well be 3 separate global functions, but logically they belong together and perhaps in the future we'll have more functions or member variables that would fit there.